### PR TITLE
Making cgroup telemetry work end to end, with new spec for metrics

### PR DESCRIPTION
**Pull request by vrdmr**
## Description
Enabling Telemetry E2E for Cgroup Perf counters for extensions.

---

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [x] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [x] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

---
This change is [Reviewable](https://reviewable.io/reviews/azure/walinuxagent/1538)

### CODE REVIEW

**Comment by narrieta** [comment]
azurelinuxagent/common/cgroups/__init__.py is empty (not needed)

File: azurelinuxagent/common/cgroup.py [hunk 1]
```diff
@@ -13,104 +13,139 @@
...
-        :param str hierarchy: Name of cgroup metric hierarchy
-        :param str file_name: Name of file within that metric hierarchy
+        :param str controller: Name of cgroup metric controller
+        :param str file_name: Name of file within that metric controller
         :return: Entire contents of the file
         :rtype: str
         """
-        if hierarchy in self.cgroups:
-            parameter_file = self._get_cgroup_file(hierarchy, file_name)
+        if controller == self.controller:
```
Comment (pgombar) [inline azurelinuxagent/common/cgroup.py]: Why do we need this check and why do we need to pass the `controller` variable instead of just using `self.controller`? I might be missing something here.

File: azurelinuxagent/common/cgroup.py [hunk 2]
```diff
@@ -152,29 +187,41 @@ def collect(self):
...
+        super(MemoryCgroup, self).__init__(extension_name, cgroup_path, controller)
+
+    def _get_memory_usage(self):
+        """
+        Collect memory.usage_in_bytes from the cgroup.
+
+        :return: Memory usage in bytes
+        :rtype: int
+        """
+        usage = self.get_parameter('memory', 'memory.usage_in_bytes')
```
Comment (pgombar) [inline azurelinuxagent/common/cgroup.py]: Is there a reason you are not using `self.controller` instead of `'memory'` like in L224?

File: azurelinuxagent/common/cgroupapi.py [hunk 3]
```diff
@@ -167,6 +168,15 @@ def _get_extension_cgroup_path(self, controller, extension_name):
 
         return os.path.join(extensions_root, cgroup_name)
 
+    def _get_extension_cgroup(self, controller, extension_name):
```
Comment (pgombar) [inline azurelinuxagent/common/cgroupapi.py]: Maybe `create` instead of `get` is a more descriptive function name.

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 4]
```diff
@@ -13,47 +13,184 @@
...
+        return [metric.average(), metric.min(), metric.max(), metric.median(), metric.count(), metric.first_polltime(),
+                metric.last_polltime()]
+
+    @staticmethod
+    def _process_cgroup_metric(cgroup_metrics):
+        current_memory_usage, max_memory_levels, current_cpu_usage = cgroup_metrics.get_metrics()
+        processed_extension = {
+            "memory": {
+                "cur_mem": CGroupsTelemetry._get_metrics_list(current_memory_usage),
+                "max_mem": CGroupsTelemetry._get_metrics_list(max_memory_levels)
```
Comment (pgombar) [inline azurelinuxagent/common/cgroupstelemetry.py]: What's the need for average, min, max, median, count... for max memory?

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 5]
```diff
@@ -13,47 +13,184 @@
...
+        CGroupsTelemetry._rlock.acquire()
+        for cgroup in CGroupsTelemetry._tracked:
+            if not cgroup.is_active():
+                CGroupsTelemetry._tracked.remove(cgroup)
+
+        CGroupsTelemetry._rlock.release()
+
+    @staticmethod
+    def cleanup():
+        CGroupsTelemetry._tracked *= 0  # emptying the list
```
Comment (pgombar) [inline azurelinuxagent/common/cgroupstelemetry.py]: This is a pretty neat way of doing this. I hope it's compatible with all our Python versions!

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 6]
```diff
@@ -13,47 +13,184 @@
...
+        if controller is "cpu":
+            self._add_cpu_usage(metric)
+        elif controller is "memory":
+            self._add_memory_usage(metric)
+
+    def get_metrics(self):
+        return self.current_memory_usage, self.max_memory_levels, self.current_cpu_usage
+
+    def clear(self):
+        self.current_memory_usage.clear()
```
Comment (pgombar) [inline azurelinuxagent/common/cgroupstelemetry.py]: `clear()` isn't supported for lists in Python <3.3

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 7]
```diff
@@ -13,47 +13,184 @@
...
+
+    def append(self, metric):
+        self.data.append(metric)
+        self.last_poll_time = dt.utcnow()
+
+    def clear(self):
+        self.data *= 0
+
+    def average(self):
+        return sum(self.data) / len(self.data)
```
Comment (pgombar) [inline azurelinuxagent/common/cgroupstelemetry.py]: We need an explicit cast to float here if we don't want to end up with an int value and use integer division.

File: tests/common/test_cgroups.py [hunk 8]
```diff
@@ -304,6 +321,129 @@ def test_limits(self):
...
+            name = "dummy_extension_{0}".format(i)
+
+            self.assertIn(name, collected_metrics)
+            self.assertIn("memory", collected_metrics[name])
+            self.assertIn("cur_mem", collected_metrics[name]["memory"])
+            self.assertIn("max_mem", collected_metrics[name]["memory"])
+            self.assertEqual(num_summarization_values, len(collected_metrics[name]["memory"]["cur_mem"]))
+            self.assertEqual(num_summarization_values, len(collected_metrics[name]["memory"]["max_mem"]))
+
+            self.assertListEqual(generate_metric_list(memory_usage_values),
```
Comment (pgombar) [inline tests/common/test_cgroups.py]: Can you add another test where you actually assert concrete values that got computed? This way we can check the validity of the computations as well. I'm mostly concerned about division.

File: azurelinuxagent/common/cgroupstelemetry.py (hunk 6 above)
Comment (vrdmr) [reply azurelinuxagent/common/cgroupstelemetry.py]: This clear calls the clear within Metric() (which internally does *= 0)

File: azurelinuxagent/common/cgroupstelemetry.py (hunk 5 above)
Comment (vrdmr) [reply azurelinuxagent/common/cgroupstelemetry.py]: Verified between Py2 and 3. :)

File: azurelinuxagent/common/cgroup.py [hunk 9]
```diff
@@ -13,104 +13,139 @@
...
+from azurelinuxagent.common import logger
 from azurelinuxagent.common.exception import CGroupsException
+from azurelinuxagent.common.future import ustr
 from azurelinuxagent.common.osutil import get_osutil
+from azurelinuxagent.common.utils import fileutil
 
-re_user_system_times = re.compile('user (\d+)\nsystem (\d+)\n')
+re_user_system_times = re.compile(r'user (\d+)\nsystem (\d+)\n')
 
 # The metric classes (Cpu, Memory, etc) can all assume that CGroups is enabled, as the CGroupTelemetry
```
Comment (narrieta) [inline azurelinuxagent/common/cgroup.py]: let's remove this comment, it is no longer applicable

File: azurelinuxagent/common/cgroup.py [hunk 10]
```diff
@@ -13,104 +13,139 @@
...
 
 
 class CGroup(object):
-    def __init__(self):
-        pass
-
-    def _get_cgroup_file(self, hierarchy, file_name):
-        return os.path.join(self.cgroups[hierarchy], file_name)
+    def __init__(self, extension_name, cgroup_path, controller_type):
+        self.name = extension_name
```
Comment (narrieta) [inline azurelinuxagent/common/cgroup.py]: what is the name used for? i think i saw it only on an error message that also includes the path... the path should be enough to identify the cgroup.

If we do need the name it should not be "extension_name", since we will also have cgroups for the agent, services, etc

File: azurelinuxagent/common/cgroup.py (hunk 1 above)
Comment (narrieta) [reply azurelinuxagent/common/cgroup.py]: +1 --- Also, in the next round of refactoring let's eliminate the controller_type, as we discussed.

File: azurelinuxagent/common/cgroup.py [hunk 11]
```diff
@@ -13,104 +13,139 @@
...
         try:
-            values = self.get_file_contents(hierarchy, parameter_name).splitlines()
+            values = self.get_file_contents(controller, parameter_name).splitlines()
             result = values[0]
         except IndexError:
-            parameter_filename = self._get_cgroup_file(hierarchy, parameter_name)
+            parameter_filename = self._get_cgroup_file(parameter_name)
+            logger.error("File {0} is empty but should not be".format(parameter_filename))
+        except CGroupsException:
+            # ignore if the file does not exist yet
```
Comment (narrieta) [inline azurelinuxagent/common/cgroup.py]: this comment is misleading, or we are catching the wrong exception... seems to me we can get CGroupsException under conditions other that the file not existing

File: azurelinuxagent/common/cgroup.py [hunk 12]
```diff
@@ -13,104 +13,139 @@
...
         Retrieve the value of a parameter from a hierarchy.
         Assumes the parameter is the sole line of the file.
 
-        :param str hierarchy: Name of cgroup metric hierarchy
+        :param str controller: Name of cgroup metric hierarchy
         :param str parameter_name: Name of file within that metric hierarchy
         :return: The first line of the file, without line terminator
         :rtype: str
         """
         result = ""
```
Comment (narrieta) [inline azurelinuxagent/common/cgroup.py]: None may be a better choice that "" for the return value

File: azurelinuxagent/common/cgroup.py [hunk 13]
```diff
@@ -13,104 +13,139 @@
...
         else:
-            raise CGroupsException("{0} subsystem not available in cgroup {1}. cgroup paths: {2}".format(
-                hierarchy, self.name, self.cgroups))
+            raise CGroupsException("Incorrect controller: {0} passed for cgroup: {1}. Cgroup path is: {2}".format(
+                controller, self.name, self.path))
 
-    def get_parameter(self, hierarchy, parameter_name):
+    def get_parameter(self, controller, parameter_name):
         """
         Retrieve the value of a parameter from a hierarchy.
```
Comment (narrieta) [inline azurelinuxagent/common/cgroup.py]: "hierarchy" -> "controller"

File: azurelinuxagent/common/cgroup.py [hunk 14]
```diff
@@ -13,104 +13,139 @@
...
+        self.controller = controller_type
 
+    def _get_cgroup_file(self, file_name):
+        return os.path.join(self.path, file_name)
 
-    def get_file_contents(self, hierarchy, file_name):
+    def get_file_contents(self, controller, file_name):
         """
-        Retrieve the value of a parameter from a hierarchy.
+        Retrieve the value of a parameter from a controller.
```
Comment (narrieta) [inline azurelinuxagent/common/cgroup.py]: >  Retrieve the value of a parameter from a controller

that is what get_parameter does... this comment should probably state that the function returns the contents of a file

File: azurelinuxagent/common/cgroup.py [hunk 15]
```diff
@@ -13,104 +13,139 @@
...
-                raise CGroupsException("Could not retrieve cgroup file {0}/{1}".format(hierarchy, file_name))
+                raise CGroupsException("Could not retrieve cgroup file {0}/{1}".format(self.path, file_name))
         else:
-            raise CGroupsException("{0} subsystem not available in cgroup {1}. cgroup paths: {2}".format(
-                hierarchy, self.name, self.cgroups))
+            raise CGroupsException("Incorrect controller: {0} passed for cgroup: {1}. Cgroup path is: {2}".format(
+                controller, self.name, self.path))
 
-    def get_parameter(self, hierarchy, parameter_name):
+    def get_parameter(self, controller, parameter_name):
```
Comment (narrieta) [inline azurelinuxagent/common/cgroup.py]: it is too since to get confused between get_parameter and get_parameters.

probably this assumption should be verified by the code
> Assumes the parameter is the sole line of the file.

or just ged rid of get_parameter and always use get_parameters

File: azurelinuxagent/common/cgroup.py [hunk 16]
```diff
@@ -13,104 +13,139 @@
...
+    def __init__(self, extension_name, cgroup_path, controller_type):
+        self.name = extension_name
+        self.path = cgroup_path
+        self.controller = controller_type
 
+    def _get_cgroup_file(self, file_name):
+        return os.path.join(self.path, file_name)
 
-    def get_file_contents(self, hierarchy, file_name):
+    def get_file_contents(self, controller, file_name):
```
Comment (narrieta) [inline azurelinuxagent/common/cgroup.py]: i think this should be private

File: azurelinuxagent/common/cgroup.py (hunk 15 above)
Comment (narrieta) [inline azurelinuxagent/common/cgroup.py]: should be private

File: azurelinuxagent/common/cgroup.py [hunk 17]
```diff
@@ -185,8 +232,16 @@ def collect(self):
...
+
+
+class CollectedMetrics(object):
+    def __init__(self, controller, metric_name, value):
+        self.controller = controller
+        self.metric_name = metric_name
+        self.value = value
 
 #
 # TODO: Do we need this code?
```
Comment (narrieta) [inline azurelinuxagent/common/cgroup.py]: I copy-pasted this code here in case we needed it. If we don't, let's remove it

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 18]
```diff
@@ -13,47 +13,184 @@
...
 
     @staticmethod
     def track_cgroup(cgroup):
         """
-        Adds the given item to the list of tracked cgroups
+        Adds the given item to the dictionary of tracked cgroups
         """
-        pass  # TODO
+        # Making track thread-safe.
+        CGroupsTelemetry._rlock.acquire()
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupstelemetry.py]: let's acquire/release the lock using try/finally

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 19]
```diff
@@ -13,47 +13,184 @@
...
+        CGroupsTelemetry._rlock.release()
 
     @staticmethod
-    def is_tracked(name):
+    def is_tracked(name, controller):
         """
         Returns true if the given item is in the list of tracked items
         """
-        pass  # TODO
+        return name in CGroupsTelemetry._tracked and controller in CGroupsTelemetry._tracked[name]
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupstelemetry.py]: should the access to _tracked be protected by the lock?

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 20]
```diff
@@ -13,47 +13,184 @@
...
 
     @staticmethod
     def stop_tracking(name):
         """
         Stop tracking the cgroups for the given name
         """
-        if CGroupsTelemetry.is_tracked(name):
-            del (CGroupsTelemetry._tracked[name])  # TODO - thread-safeness; review implementation
+        # Pop is atomic and thread-safe.
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupstelemetry.py]: > Pop is atomic and thread-safe.

What about append?

If one of them is not thread safe then probably we should protect all of them regardless of thread-safety

File: azurelinuxagent/common/cgroup.py (hunk 2 above)
Comment (vrdmr) [reply azurelinuxagent/common/cgroup.py]: Removed the controller itself. No need to handle this hot-potato. :)

File: azurelinuxagent/common/cgroupstelemetry.py (hunk 4 above)
Comment (vrdmr) [reply azurelinuxagent/common/cgroupstelemetry.py]: First - Implementation-wise it was easier to manage all the metrics as of type Metric. I am thinking of simplifying it with the next round of change, but this makes it simple.

```
class CgroupMetrics(object):
    def __init__(self):
        self.current_memory_usage = Metric()
        self.max_memory_levels = Metric()
        self.current_cpu_usage = Metric()
```

Also, the Max of the Max list can still give the absolute Max, and its a value which can change as well.

File: tests/common/test_cgroups.py (hunk 8 above)
Comment (vrdmr) [reply tests/common/test_cgroups.py]: Yup. I am adding that in the test_extension test, which will do the entire extension_handler.run() and then do a poll_tracked_list. That unittest will go in another PR.

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 21]
```diff
@@ -13,47 +13,184 @@
...
+            "cpu": {
+                "cur_cpu": CGroupsTelemetry._get_metrics_list(current_cpu_usage)
+            }
+        }
+        return processed_extension
 
     @staticmethod
     def metrics_hierarchies():
-        return CGroupsTelemetry._hierarchies
+        pass  # return CGroupsTelemetry._hierarchies
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupstelemetry.py]: left-over?

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 22]
```diff
@@ -13,47 +13,184 @@
...
+                CGroupsTelemetry._tracked.remove(cgroup)
+
+        CGroupsTelemetry._rlock.release()
+
+    @staticmethod
+    def prune_all_tracked():
+        CGroupsTelemetry._rlock.acquire()
+        for cgroup in CGroupsTelemetry._tracked:
+            if not cgroup.is_active():
+                CGroupsTelemetry._tracked.remove(cgroup)
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupstelemetry.py]: can we log the cgroups we are pruning?

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 23]
```diff
@@ -13,47 +13,184 @@
...
     @staticmethod
     def track_cgroup(cgroup):
         """
-        Adds the given item to the list of tracked cgroups
+        Adds the given item to the dictionary of tracked cgroups
         """
-        pass  # TODO
+        # Making track thread-safe.
+        CGroupsTelemetry._rlock.acquire()
+        CGroupsTelemetry._tracked.append(cgroup)
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupstelemetry.py]: can we log the cgroup we are beginning to track?

File: tests/common/test_cgroups.py [hunk 24]
```diff
@@ -69,6 +67,25 @@ def i_am_root():
...
+
+def generate_metric_list(lst):
+    return [sum(lst) / len(lst),
+            min(lst),
+            max(lst),
+            median(lst),
+            len(lst)]
+
+
 @skip_if_predicate_false(lambda: False, "TODO: Need unit tests")
```
Comment (narrieta) [inline tests/common/test_cgroups.py]: I disabled the old tests since we were rewriting this code... could you check whether we need any of those tests and remove them if not?

File: azurelinuxagent/ga/monitor.py [hunk 25]
```diff
@@ -253,13 +262,14 @@ def collect_and_send_events(self):
...
-                        MonitorHandler.CGROUP_TELEMETRY_PERIOD,
+                        MonitorHandler.CGROUP_TELEMETRY_POLLING_PERIOD,
+                        MonitorHandler.CGROUP_TELEMETRY_REPORTING_PERIOD,
                         MonitorHandler.EVENT_COLLECTION_PERIOD,
                         MonitorHandler.HOST_PLUGIN_HEARTBEAT_PERIOD,
                         MonitorHandler.IMDS_HEARTBEAT_PERIOD).seconds
         while self.should_run:
             self.send_telemetry_heartbeat()
-            self.send_cgroup_telemetry()
+            self.send_extension_metrics_telemetry()
```
Comment (narrieta) [inline azurelinuxagent/ga/monitor.py]: minor: name needs to be a little more generic, since this is for any cgroups (extensions, agent, services)

File: azurelinuxagent/ga/monitor.py [hunk 26]
```diff
@@ -399,54 +409,38 @@ def send_telemetry_heartbeat(self):
...
-                                          is_success=True,
-                                          message=msg,
-                                          log_event=True)
-
-            except Exception as e:
-                logger.warn("Monitor: failed to collect cgroups performance metrics: {0}", ustr(e))
-                logger.verbose(traceback.format_exc())
-
-            self.last_cgroup_telemetry = datetime.datetime.utcnow()
+    def send_extension_metrics_telemetry(self):
```
Comment (narrieta) [inline azurelinuxagent/ga/monitor.py]: minor: can we separate this into two functions (poll and report) and call both from the main loop (makes it a little clearer that they are separate operations with different periods)

File: azurelinuxagent/common/cgroup.py (hunk 10 above)
Comment (vrdmr) [reply azurelinuxagent/common/cgroup.py]: We need the name in the Cgroup metrics. Yes, I'll change the parameter of the `__init__` to `name` rather than `extension_name`

File: azurelinuxagent/common/cgroupconfigurator.py [hunk 27]
```diff
@@ -166,9 +167,12 @@ def start_extension_command(self, extension_name, command, cwd, env, stdout, std
                 def track_cgroups():
                     cgroups = self._cgroups_api.get_extension_cgroups(extension_name)
 
-                    # TODO: Add to tracking list
+                    for cgroup in cgroups:
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupconfigurator.py]: we also need to track the agent in create_agent_cgroups

Comment (narrieta) [inline azurelinuxagent/common/cgroupconfigurator.py]: there is a TODO in remove_extension_cgroups that probably it is not needed anymore given the way you are pruning the tracking list (I like that)... could you remove the TODO?

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 28]
```diff
@@ -13,47 +13,184 @@
...
+    def clear(self):
+        self.current_memory_usage.clear()
+        self.max_memory_levels.clear()
+        self.current_cpu_usage.clear()
+
+
+class Metric(object):
+    def __init__(self):
+        self.data = []
+        self.first_poll_time = dt.utcnow()
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply azurelinuxagent/common/cgroupstelemetry.py]: Refactored all the three (data/polltimes) to be private.

File: azurelinuxagent/ga/monitor.py (hunk 25 above)
Comment (vrdmr) [reply azurelinuxagent/ga/monitor.py]: The `extension` in `send_extension_metrics_telemetry` was targetted more on the extension as a package (which could depend on other things on the box as well like services), rather than only focusing on extensions runtime. It could in the future send not only cgroups but extension status, python package information, etc.

File: azurelinuxagent/common/cgroupconfigurator.py (hunk 27 above)
Comment (vrdmr) [reply azurelinuxagent/common/cgroupconfigurator.py]: Good catch. I will add it here.

File: azurelinuxagent/ga/monitor.py (hunk 26 above)
Comment (vrdmr) [reply azurelinuxagent/ga/monitor.py]: When you say - 

> call both from the main loop

With the main loop, do you mean to add another method call only for polling, in monitor.py#270? I wanted the method to be self-contained with the task of handling all the extension_metrics related operations.
```
while self.should_run:
    ...
    self.send_extension_metrics_telemetry()
    ...
```

I have tried to simplify the `send_extension_metrics_telemetry` method further. Let me know if that works. We can discuss this as well.

File: azurelinuxagent/common/cgroupconfigurator.py [hunk 29]
```diff
@@ -97,8 +98,8 @@ def __impl():
                 cgroups = self._cgroups_api.create_agent_cgroups()
 
                 if track_cgroups:
-                    # TODO: Add to tracking list
-                    pass
+                    for cgroup in cgroups:
+                        CGroupsTelemetry.track_cgroup(cgroup)
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupconfigurator.py]: CGroupApi.create_agent_cgroups needs to be updated to return CGroups instead of directory paths

File: azurelinuxagent/ga/monitor.py (hunk 25 above)
Comment (narrieta) [reply azurelinuxagent/ga/monitor.py]: if we add more data, it is likely the reporting period will be different. I would make this specific to resource metrics

File: azurelinuxagent/ga/monitor.py (hunk 26 above)
Comment (narrieta) [reply azurelinuxagent/ga/monitor.py]: Since it is 2 different operations (collect and send) with different periods I think having 2 calls from the main loop makes the intention clearer

File: tests/common/test_cgroups.py [hunk 30]
```diff
@@ -304,6 +321,133 @@ def test_limits(self):
         self.assert_limits(ext_name="normal_extension", expected_cpu_limit=40, exception_raised=True)
 
 
+class TestCGroupsTelemetry(AgentTestCase):
```
Comment (narrieta) [inline tests/common/test_cgroups.py]: let's move this to test_cgroupstelemetry.py

File: azurelinuxagent/common/cgroup.py [hunk 31]
```diff
@@ -13,104 +13,108 @@
...
-# class is very careful not to call them if CGroups isn't enabled. Any tests should be disabled if the osutil
-# is_cgroups_support() method returns false.
+re_user_system_times = re.compile(r'user (\d+)\nsystem (\d+)\n')
 
 
 class CGroup(object):
-    def __init__(self):
-        pass
+    def __init__(self, name, cgroup_path, controller_type):
+        self.name = name
```
Comment (narrieta) [inline azurelinuxagent/common/cgroup.py]: why do cgroups need a name? the path is already unique

File: tests/common/test_cgroups.py [hunk 32]
```diff
@@ -304,6 +321,133 @@ def test_limits(self):
...
+                        patch_get_cpu_percent.return_value = 30
+                        patch_get_memory_usage.return_value = 209715200  # example 200 MB
+                        patch_get_memory_max_usage.return_value = 471859200  # example 450 MB
+
+                        CGroupsTelemetry.poll_all_tracked()
+
+                        self.assertEqual(CGroupsTelemetry._cgroup_metrics.__len__(), num_extensions)
+                        for cgroup_name, cgroup_metric in CGroupsTelemetry._cgroup_metrics.items():
+                            current_memory_usage, max_memory_levels, current_cpu_usage = cgroup_metric.get_metrics()
+                            self.assertEqual(len(current_memory_usage._data), 1)
```
Comment (narrieta) [inline tests/common/test_cgroups.py]: should we check the values too? (not only the length)

File: azurelinuxagent/ga/monitor.py [hunk 33]
```diff
@@ -399,54 +409,39 @@ def send_telemetry_heartbeat(self):
...
+
+        if not self.last_cgroup_polling_telemetry:
+            self.last_cgroup_polling_telemetry = time_now
+
+        if not self.last_cgroup_report_telemetry:
+            self.last_cgroup_report_telemetry = time_now
+
+        if time_now >= (self.last_cgroup_polling_telemetry +
+                        MonitorHandler.CGROUP_TELEMETRY_POLLING_PERIOD):
+            CGroupsTelemetry.poll_all_tracked()
```
Comment (narrieta) [inline azurelinuxagent/ga/monitor.py]: The names for poll_all_tracked and collect_all_tracked are too similar... it is hard to see at a glance what they are doing. 

I think CGROUP_TELEMETRY_POLLING_PERIOD and CGROUP_TELEMETRY_REPORTING_PERIOD express very clearly their intent. Can we rename to include "reporting" somewhere?

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 34]
```diff
@@ -13,47 +13,188 @@
...
-        pass  # TODO
+        collected_metrics = {}
+        for name, cgroup_metrics in CGroupsTelemetry._cgroup_metrics.items():
+            collected_metrics[name] = CGroupsTelemetry._process_cgroup_metric(cgroup_metrics)
+            cgroup_metrics.clear()
+
+            if cgroup_metrics.marked_for_delete:
+                # Clear the _cgroup_metrics to make sure we don't send stale entries,
+                # only after sending its content first
+                CGroupsTelemetry._cgroup_metrics.pop(name, None)
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupstelemetry.py]: > only after sending its content first

Yes, that's what it should do... but we are clearing before sending

File: tests/common/test_cgroups.py (hunk 30 above)
Comment (narrieta) [inline tests/common/test_cgroups.py]: The current tests cover poll and collect... we need tests for the rest of the methods

File: azurelinuxagent/common/cgroupstelemetry.py (hunk 18 above)
Comment (vrdmr) [reply azurelinuxagent/common/cgroupstelemetry.py]: Changed this to use `with` as suggested in [here](https://stackoverflow.com/a/25704099). This takes care of releasing the lock if we hit an issue.

Thanks for catching this.

File: azurelinuxagent/common/cgroupstelemetry.py (hunk 34 above)
Comment (vrdmr) [reply azurelinuxagent/common/cgroupstelemetry.py]: Its metrics are captured in line 81 - `collected_metrics[name] = CGroupsTelemetry._process_cgroup_metric(cgroup_metrics)`, and we return `collected_metrics` from `collect_all_tracked()`

File: azurelinuxagent/common/cgroup.py (hunk 31 above)
Comment (vrdmr) [reply azurelinuxagent/common/cgroup.py]: The Name helps me to create a new cgroup_metric of the same name, which is used in the json metric as well.

Example: in `azurelinuxagent/common/cgroupstelemetry.py:96`
`CGroupsTelemetry._cgroup_metrics[cgroup.name] = CgroupMetrics()`

File: azurelinuxagent/common/cgroup.py [hunk 35]
```diff
@@ -152,29 +156,41 @@ def collect(self):
...
-    def get_memory_usage(self):
+    def _get_memory_usage(self):
         """
         Collect memory.usage_in_bytes from the cgroup.
 
         :return: Memory usage in bytes
         :rtype: int
         """
-        usage = self.cgt.cgroup.get_parameter('memory', 'memory.usage_in_bytes')
+        usage = self.get_parameters('memory.usage_in_bytes', first_line_only=True)
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply azurelinuxagent/common/cgroup.py]: This particular file contains only one value and the `first_line_only` in `get_parameters()` is only to get that value. I could do [0] by the calling function, but that brings some ambiguity.

File: azurelinuxagent/common/cgroup.py (hunk 11 above)
Comment (vrdmr) [reply azurelinuxagent/common/cgroup.py]: Returning None here and handling it upstream.

File: azurelinuxagent/common/cgroup.py [hunk 36]
```diff
@@ -13,104 +13,108 @@
...
+re_user_system_times = re.compile(r'user (\d+)\nsystem (\d+)\n')
 
 
 class CGroup(object):
-    def __init__(self):
-        pass
+    def __init__(self, name, cgroup_path, controller_type):
+        self.name = name
+        self.path = cgroup_path
+        self.controller = controller_type
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply azurelinuxagent/common/cgroup.py]: It is used inside the CGroupMetrics - The structure of the CGroupMetrics will be changed with the next PR, and controller will be removed then.

```
for cgroup in CGroupsTelemetry._tracked:
    if cgroup.name not in CGroupsTelemetry._cgroup_metrics:
        CGroupsTelemetry._cgroup_metrics[cgroup.name] = CgroupMetrics()
    CGroupsTelemetry._cgroup_metrics[cgroup.name].add_new_data(cgroup.controller, cgroup.collect())
```

[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/common/cgroup.py]: cool

File: azurelinuxagent/common/cgroup.py (hunk 31 above)
Comment (narrieta) [reply azurelinuxagent/common/cgroup.py]: still, you could use the path as key instead of the name... the path is guaranteed to be unique... as we add new cgroups to track services, cron jobs and what not we will need to come up with a way to name these and to ensure the name remains unique... path does not have that issue.

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 37]
```diff
@@ -13,47 +13,180 @@
...
 
     @staticmethod
-    def is_tracked(name):
+    def is_tracked(name, controller):
         """
         Returns true if the given item is in the list of tracked items
+        O(n) operation. But limited to few cgroup objects we have.
         """
-        pass  # TODO
+        for cgroup in CGroupsTelemetry._tracked:
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupstelemetry.py]: You need to hold the lock while iterating thru _tracked

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 38]
```diff
@@ -13,47 +13,180 @@
...
 
     @staticmethod
-    def stop_tracking(name):
+    def stop_tracking(cgroup):
         """
         Stop tracking the cgroups for the given name
         """
-        if CGroupsTelemetry.is_tracked(name):
-            del (CGroupsTelemetry._tracked[name])  # TODO - thread-safeness; review implementation
+        CGroupsTelemetry._tracked.remove(cgroup)
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupstelemetry.py]: please check that all uses of _tracked are protected by the lock

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 39]
```diff
@@ -13,47 +13,180 @@
...
+        with CGroupsTelemetry._rlock:
+            for cgroup in CGroupsTelemetry._tracked[:]:
+                if not cgroup.is_active():
+                    CGroupsTelemetry.stop_tracking(cgroup)
+
+    @staticmethod
+    def cleanup():
+        CGroupsTelemetry._tracked *= 0  # emptying the list
+
+    def __init__(self):
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupstelemetry.py]: leftover?

File: tests/common/test_cgroupstelemetry.py [hunk 40]
```diff
@@ -162,3 +160,81 @@ def test_telemetry_calculations(self, *args):
...
+        num_extensions = 5
+        num_controllers = 2
+        for i in range(num_extensions):
+            dummy_cpu_cgroup = CGroup.create("dummy_path", "cpu", "dummy_extension_{0}".format(i))
+            CGroupsTelemetry.track_cgroup(dummy_cpu_cgroup)
+
+            dummy_memory_cgroup = CGroup.create("dummy_path", "memory", "dummy_extension_{0}".format(i))
+            CGroupsTelemetry.track_cgroup(dummy_memory_cgroup)
+
+        self.assertEqual(num_extensions * num_controllers, len(CGroupsTelemetry._tracked))
```
Comment (narrieta) [inline tests/common/test_cgroupstelemetry.py]: a better assertion could be that is_tracked returns true for all the items you added (that also has the advantage of testing via the public interface, rather than reaching into _tracked)

File: tests/common/test_cgroupstelemetry.py [hunk 41]
```diff
@@ -162,3 +160,81 @@ def test_telemetry_calculations(self, *args):
...
+
+        CGroupsTelemetry.prune_all_tracked()
+        self.assertEqual(0, len(CGroupsTelemetry._tracked))
+
+        for i in CGroupsTelemetry._tracked:
```
Comment (narrieta) [inline tests/common/test_cgroupstelemetry.py]: why is this needed?

File: azurelinuxagent/common/cgroupstelemetry.py (hunk 37 above)
Comment (vrdmr) [reply azurelinuxagent/common/cgroupstelemetry.py]: This would cause locking issues in the `poll_all_tracked`. The `poll_all_tracked` has to lock to make sure it is polling data on the latest stable list. The `CGroupsTelemetry.stop_tracking(cgroup)` would hit the lock acquisition failure.

Current implementation:
```
def poll_all_tracked():
    with CGroupsTelemetry._rlock:
        for cgroup in CGroupsTelemetry._tracked[:]:
            if cgroup.name not in CGroupsTelemetry._cgroup_metrics:
                CGroupsTelemetry._cgroup_metrics[cgroup.name] = CgroupMetrics()
            CGroupsTelemetry._cgroup_metrics[cgroup.name].add_new_data(cgroup.controller, cgroup.collect())

            if not cgroup.is_active():
                CGroupsTelemetry.stop_tracking(cgroup)
                CGroupsTelemetry._cgroup_metrics[cgroup.name].marked_for_delete = True
```

If you want the locking to only happen in primitive operations such as `stop_tracking` and `track_cgroup`, It will still work.

File: tests/common/test_cgroupstelemetry.py (hunk 41 above)
Comment (vrdmr) [reply tests/common/test_cgroupstelemetry.py]: leftover. Cleaned
